networking/src/Node/Internal.hs

```diff
@@ -111,6 +115,8 @@ data NodeState peerData m = NodeState {
     , _nodeStateClosed             :: !Bool
       -- ^ Indicates whether the Node has been closed and is no longer capable
       --   of establishing or accepting connections (its EndPoint is closed).
+    , _nodeStateTimeouts           :: !(Map NodeId Microsecond)
```

Comment:
How about `Map NodeId [Microsecond]`. The head of this list (if nonempty) is the next delay. If the list is empty, it means all delays have run their course, so the connection attempt should fail. This will make `connectToPeer` logic simpler, as it won't need the partial `last :: [a] -> a` function.

Reply:
The `last` is not partial here because it consumes a non empty list.  What if all attempt failed, but after a while the failing node accepts connections back again.  In this case you will not reach to it (unless you restart the node).